neqo-http3/src/connection_client.rs

```diff
@@ -6521,4 +6612,32 @@ mod tests {
         client.process(out.dgram(), now());
         assert_closed(&client, &Error::HttpGeneralProtocol);
     }
+
+    #[test]
+    fn priority_update_during_full_buffer() {
+        // set a lower max_data on the server side to restrict the data the client can send
+        let (mut client, mut server) =
+            connect_with_connection_parameters(ConnectionParameters::default().max_data(1200));
+        let request_stream_id = make_request_and_exchange_pkts(&mut client, &mut server, false);
+        let data_writable = |e| matches!(e, Http3ClientEvent::DataWritable { .. });
+        assert!(client.events().any(data_writable));
+        // Send a lot of data
+        client
+            .send_request_body(request_stream_id, &[0; 3000])
+            .unwrap();
+
+        // now change priority
+        assert!(client
+            .priority_update(request_stream_id, Priority::new(6, false))
+            .unwrap());
+
```

Comment:
Can we extend this test? Currently it only tests that there is not loop.
Can we  do the following:
- after this line, let the client and server exchange packets and make sure that the sever does not receive a priority update (to check this it will be difficult because the server in this tests does not have http layer. You can read data from the control stream(I think the control stream will have stream_id 2) and check thaa it does not have any data or maybe only a SETTINGS frame).
- read stream data on he server side -> this will cause the server to queue a max_data update frame,
-  the client and server exchange packets again so that the client receives the max_data update frame,
- now server will receive a priority update (here as well you need to read the control stream and check that i contains a priority frame).

Reply:
This doesn't work as I expected for me. The priority_update seems to get send (in the second packet I think):

doing the http request and priority update before the following packet exchange:

1. client -> server (partial http request, but priority_update doesn't fit)
2. server -> client (ack I guess)
3. client -> server (priority_update)

afterwards, reading the control_stream returns the priority update

I think this is because each stream has a separate flow control https://http3-explained.haxx.se/en/quic/quic-streams#flow-control
And I'm just restricting the flow control, client control stream has still free buffer to fill.

Reply:
There is per-stream flow control and per-complete quic session. Max_data should change the session's flow control.
Maybe something is freeing flow control credits, but I cannot say what without seeing the test....
when I think about it it is possible that enough flow control  credits is freed by reading SETTINGS frame and the request HEADER frame. And in addition the client sends MAX_DATA_BLOCKED frame that will cause server to free credits immediately. (my original thinking was wrong about this)

So please write the test as you have describe it above. It is fine as well.
In the step 2 (from your description) add a check that max_data frame has been sent:
```
let md_before = server.stats().frame_tx.max_data;
let out = server.process(..);
assert_eq!(server.stats().frame_tx.max_data, md_before + 1);
```